Binary sensors follow up

marcelveldt opened the pull request:
The is the "deluxe" follow-up on the previous PR to create more sensible/userfriendly sensors based on values in the Notification Command Class.
Based on discussions with @kpine and @cgarwood I've created a mapping from the Z-Wave listvalues into Home Assistant binary sensors.

Also, this will change the default naming to "[short device name]: [Notification Value label] - [Notification Value Listitem label]"

marcelveldt committed:
userfriendly binary sensors

Follow-up on the userfriendly binary sensors to make it feature complete
Handle all possible sensors thanks to a mapping table for the entire notification CommandClass values.

marcelveldt committed:
formatting and linting

marcelveldt committed:
fix test

MartinHjelmare commented on custom_components/zwave_mqtt/entity.py:
```diff
@@ -147,17 +147,24 @@ def __init__(self, values):
         self.async_write_ha_state()
 
+    @callback
+    def values_updated(self):
```
Is this needed because we want to take the same action when either `value_added` or `value_changed` is called?

marcelveldt replied:
Yeah, but I wanted to make it more generic (and fail safe) to prevent common mistakes such as forgetting to call the original super() method with common logic. This leaves the common value parsing in the base entity class.

marcelveldt replied:
And yes, in this case we want to take the same action for value_added and value_changed

MartinHjelmare replied:
Can we document more explicitly when a platform should use the `values_updated` callback and when it should use the other two `value_` callbacks? It's not clear to me when `values_updated` should be used and when `value_changed` should be used eg. Why do we name it `values` instead of `value`?

marcelveldt replied:
Option 1:
Just leave it with value_changed / value_added and let each platform override this method if they see fit.

Option 2:
Make the value_added / value_changed methods private. 
Advertise "values_updated" or "value_updated" as the one to use within a platform if you like to be informed about the underlying data collection being updated.

MartinHjelmare replied:
I think we should make it as strict as possible. So if option 2 gives enough flexibility, let's go for that.

MartinHjelmare replied:
Although, since we're inheriting from this entity, subclasses will be able to override private methods too. So we should document if we don't want that.

marcelveldt replied:
My personal preference goes to 2 with clear comments in the private methods. This will make the code easier to maintain/read in the longer end I think. You already see some duplicating going on in the platforms.

MartinHjelmare replied:
Sounds good!

marcelveldt replied:
Preference for a name ?
"value_updated"
"values_updated"
"entity_updated"
"entity_values_updated"
"on_value_update"

MartinHjelmare replied:
I like `on_value_update`. That is more distinct from the other names we already have while keeping `value` in singular mode.

marcelveldt replied:
OK, I've adjusted this. The only platform affected (atm) was the light platform so I kept the change here instead of separate PR.

kpine commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -21,26 +22,261 @@
+    {
+        # Index 1: Smoke Alarm - Value Id's 1 and 2
+        NOTIFICATION_TYPE: NOTIFICATION_SMOKE_ALARM,
+        NOTIFICATION_VALUES: [1, 2],
```
So does this result in two different binary sensors?

1. `binary_sensor.<device>_smoke_alarm_smoke_detected_at_location`
2. `binary_sensor.<device>_smoke_alarm_smoke_detected_at_unknown_location`

But then I noticed you are stripping from `at` or `on` and after, which would cause both of these to have the same name. Are you assuming a device would not support both values, or what would happen if it did?

For these values in particular, I think it would make sense for them to always be the same sensor. The location event data should be an attribute?

marcelveldt replied:
As far as I understand the specs (and seen it on my own devices) you will not have both sensors at the same time present. Also I have never seen the location filled.
Do you say otherwise ?

kpine replied:
Nope, I can't say either way. I don't see anything that says they are mutually exclusive though. I only have an old smoke alarm that is V1, so it's much worse.

marcelveldt replied:
OK, I'll put in some comments in the code. If someone actually owns a device that has both values present he will get 2 binary sensors with the same name and report that. In that case we can extend the code. I have a strong feeling that is never going to happen :-)

marcelveldt replied:
Added some comments in the code.

kpine commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -52,27 +288,24 @@ def async_add_binary_sensor(values):
-                    # Motion detected
+            for list_value in values.primary.value["List"]:
+                # check if we have a mapping for this value
+                for item in NOTIFICATION_SENSORS:
+                    if item[NOTIFICATION_TYPE] != values.primary.index:
+                        continue
```
Why a list, instead of a dictionary with the notification type as key (then list of sensors)?

marcelveldt replied:
Flexibility, some notification types have multiple sensors.

kpine commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -21,26 +22,261 @@
+    {
+        # Index 4: Heat - Value Id's 1, 2, 5, 6 (heat/underheat)
+        NOTIFICATION_TYPE: NOTIFICATION_HEAT,
+        NOTIFICATION_VALUES: [1, 2, 5, 6],
```
Same question, are these four separate binary sensors? If so, I'm not so sure about that, especially when you consider the clear logic that has been added.

![image](https://user-images.githubusercontent.com/791794/80450491-7b3a0400-88d6-11ea-96f9-ed4e9dec1a48.png)

These appear to be defined as a single state variable. You could argue whether this is a binary_sensor or not, but if it is, I think it would be only two sensors. A single "overheat_detected" and another "underheat_detected". How can something be both overheated and underheated at the same time?

marcelveldt replied:
These will be separate sensors. A sensor will be created for each value in the list if present.

kpine commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -21,26 +22,261 @@
+    {
+        # Index 8: Power management - All Value Id's
+        NOTIFICATION_TYPE: NOTIFICATION_POWER_MANAGEMENT,
+        NOTIFICATION_VALUES: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
```
Another comment, if these are individual binary sensors...

Battery power is actually a multi-state variable.

![image](https://user-images.githubusercontent.com/791794/80450907-6b6eef80-88d7-11ea-81a4-e3f4c2a64198.png)

You could represent them as individual binary sensors, but the same Clear logic would not apply, as these sensors could not all be on at the same time. Here is an illustration:

![image](https://user-images.githubusercontent.com/791794/80450863-51cda800-88d7-11ea-8761-fd83de78222f.png)

The battery level status could not be both "fully charged" and "charge battery now" at the same time. You'd either need a normal sensor, represented with text states (yuck?), or individual binary sensors that are mutually exclusively on.

marcelveldt replied:
I have never seen a device using this and thus impossible for me to test nor code.
If you give me an example how this looks like in the MQTT messages I'd be happy to split them up. For now I consider this as an edge case and ignore it.

kpine replied:
Well, I don't have a device with these properties, but if OZW were following the spec then the MQTT messages would arrive in order of the timing diagram.
1. mqtt sends 0x0D - HA sets "battery is fully charged" to On
2. mqtt sends 0x00 - HA sets all power sensors to Off
3. mqtt sends 0x0e - HA sets "charge battery soon" sensor to On
4. mqtt sends 0x0f - HA sets "charge battery now" sensor to On
5. mqtt sends 0x0f - no state change in HA
6. mqtt sends 0x00 - HA sets all power sensors to Off
7. mqtt sends 0x0d - HA sets "battery is fully charged" to On

After step 4, both "charge battery soon" and "charge battery now" are On, which is not quite right.

marcelveldt replied:
You are correct but

1) I still have no idea how this is presented in the mqtt data structure, will there be another list in the value for example ?

2) I still think this is an edge case as I have never seen any devices with these values. My focus was on the values that can be easily presented as a binary sensor. The fallback is in the sensor with the raw sensor data. As soon as the first user pops up with this need we can extend the code but adding it in now feels beyond MVP scope.

What I will do is add some comments in the code for and ignore the values that can no be properly translated into a boolean

kpine replied:
They are just [List values](https://github.com/OpenZWave/open-zwave/blob/5220be772970b922fd6304f9d347d025a6511e7a/config/NotificationCCTypes.xml#L306-L313) like all of the others.
```xml
		<AlarmEvent id="12" name="Battery Charging">
		</AlarmEvent>
		<AlarmEvent id="13" name="Battery Charged">
		</AlarmEvent>
		<AlarmEvent id="14" name="Battery Low">
		</AlarmEvent>
		<AlarmEvent id="15" name="Battery Critical">
		</AlarmEvent>
```
I would expect that to translate to (this condensed value):
```json
{
    "Label": "Power Management",
    "Value": {
        "List": [
            {
                "Value": 0,
                "Label": "Clear"
            },
            {
                "Value": 12,
                "Label": "Battery Charging"
            }
            {
                "Value": 13,
                "Label": "Battery Charged"
            }
            ...
        ],
        "Selected": "Battery Charged",
        "Selected_id": 13
    },
    "Type": "List",
    "CommandClass": "COMMAND_CLASS_NOTIFICATION",
    "Index": 8,
}
```
The only difference is that each of these represent a value for a single variable, so they cannot all be true at the same time. Sure, you can represent them each as binary sensors, but only one can be on at a time.

My main point was that this particular set of values is a bit more complicated than the others.

marcelveldt replied:
Sorry, I mixed up my comments :-) I was talking about the case where the value itself has another list underneath. In this case it's just a regular list indeed. 

@Fishwaldo when the state moved from "charging" to "charged" will there be a "clear" event in between ?

marcelveldt replied:
Already solved.

kpine commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -21,26 +22,261 @@
+    },
+    {
+        # Index 15: Water valve
+        # ignore non-boolean values
```
Wouldn't "Valve current alarm" be an on or off value, with extra attributes for the event data? Or are you thinking the event data reflects a non-binary sensor value?

marcelveldt replied:
See my previous comment. I have no clue how this looks like in the data structure.

kpine replied:
My confusion was the comment "non-boolean values". They appear to be boolean (is there an alarm, yes or no), with extra event data to describe the alarm.

If you are ignoring those because you don't know what the event data is, then that's fine.

marcelveldt replied:
Yes, that is correct. I do not know the data structure that will be send to MQTT for these kind of values.

kpine commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -110,56 +343,40 @@ def entity_registry_enabled_default(self) -> bool:
+        """Call when the underlying values are added or updated."""
+        # Entity state only returns false if clear is reported
+        # So we do not directly map the value but wait for the value change
+        if self.values.primary.value["Selected_id"] == self._list_value:
+            self._state = True
+        elif self.values.primary.value["Selected_id"] == 0:
+            self._state = False
```
Just to repeat what I said above, for most of these this is true, but not all, such as the battery status or heat alarm heat sensor status, that have more than one state. So this might be a little simplistic for some state variables.

marcelveldt replied:
This is exactly what handles that. It will only reset if a clear event has been seen.

kpine replied:
Does this still work for doors then, or other sensors that don't support Idle (Clear)? A door/window open state is either 22 or 23, never 0. It follows this:
![image](https://user-images.githubusercontent.com/791794/80461585-5c456d00-88ea-11ea-9bfc-245f9775bfae.png)

marcelveldt replied:
it's unclear to me. What I understand is that a "previous event cleared" will be sent on the transition. Let's wait for a reply by @Fishwaldo about this.

If the clear event is not sent, the code is going to be a lot more complex, as we need to handle each specific situation.

kpine replied:
That is not always true, such as the example I just posted. Not all variables support idle.

kpine replied:
If you have not already, take a look at [SDS13713 Notification Command Class, list of assigned Notifications](https://www.silabs.com/documents/login/miscellaneous/SDS13713-Notification-Command-Class.xlsx).

Access control "door state" state variable is defined as having the values 0x16 (open) or 0x17 (closed), and Idle is not applicable.

![image](https://user-images.githubusercontent.com/791794/80468459-0c6ba380-88f4-11ea-9b7d-a57210611de1.png)

marcelveldt replied:
Clear is not meant as idle but as event to indicate the previous event cleared. So if we spot the event "door open" we set the binary_sensor ON. Only when "clear comes along we set it to OFF again.
My assumption is that OZW sends "door open" --> "clear" --> "door closed"

kpine replied:
The screen shots I've been posting are from the spec. From [OZW](https://github.com/OpenZWave/open-zwave/blob/5220be772970b922fd6304f9d347d025a6511e7a/config/NotificationCCTypes.xml#L154-L157):
```xml
	<AlarmType id="6" name="Access Control">
		<AlarmEvent id="0" name="Clear">
			<AlarmEventParam id="256" type="byte" name="Previous Event Cleared" />
		</AlarmEvent>
```
"Clear" is what OZW calls the Idle State. The value is 0x00 and the event parameter is the value of the variable that went Idle, aka "Previous Event Cleared", as shown in the screen shot.

As in the screen shot, the "Door state" variable is either open or closed. Idle is not applicable as indicated by "N/A".

Anyways, time for me to head off, I'm sure it will get sorted. 👍

marcelveldt replied:
omg, that would be horrible as it will require specific code for each situation. 
Anyways, I might have an idea to fix it in a generic way.

marcelveldt replied:
Thanks for your help on this

marcelveldt replied:
Solved. We can now handle the situation that the sensor is reset as soon as a specific other value is reached (the doorsensor example), the situation that a value resets as soon as just another value is send and the more regular "value resets on clear"

kpine reviewed (commented):
Looking better! Support for these sensors will really set this new z-wave integration apart from the old one.

marcelveldt committed:
mark value_added and value_changed as private

introduce new public method on_value_update which can be used by platforms to be notified of changes in the underlying values collection

marcelveldt committed:
final fix

Not for every event/value there will be a clear state/event reached, account for that.
Add some comments about some decisions.

marcelveldt committed:
fix light init

MartinHjelmare commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -52,27 +284,27 @@ def async_add_binary_sensor(values):
+                            values,
+                            list_value["Value"],
+                            # optional values
+                            item.get(NOTIFICATION_DEVICE_CLASS, None),
```
`None` is the default value returned if the key is missing in the dict.
```suggestion
                            item.get(NOTIFICATION_DEVICE_CLASS),
```

marcelveldt committed:
Update custom_components/zwave_mqtt/binary_sensor.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Looks great!

marcelveldt commented:
@MartinHjelmare this PR should be finished now. Please squash and merge once you're OK.
